File: regions/system/library/go/building-blocks/inmemory_secretstore_test.go
package buildingblocks

import (
	"context"
	"testing"
)

func TestInMemorySecretStore_InitAndStatus(t *testing.T) {
	s := NewInMemorySecretStore()
	ctx := context.Background()

	if s.Status(ctx) != StatusUninitialized {
		t.Errorf("expected StatusUninitialized, got %s", s.Status(ctx))
	}
	if err := s.Init(ctx, Metadata{}); err != nil {
		t.Fatalf("Init failed: %v", err)
	}
	if s.Status(ctx) != StatusReady {
		t.Errorf("expected StatusReady, got %s", s.Status(ctx))
	}
}

func TestInMemorySecretStore_Name(t *testing.T) {
	s := NewInMemorySecretStore()
	if s.Name() != "inmemory-secretstore" {
		t.Errorf("unexpected Name: %q", s.Name())
	}
	if s.Version() != "1.0.0" {
		t.Errorf("unexpected Version: %q", s.Version())
	}
}

func TestInMemorySecretStore_GetPut(t *testing.T) {
	s := NewInMemorySecretStore()
	ctx := context.Background()
	_ = s.Init(ctx, Metadata{})

	s.Put("db-password", "secret123")

	got, err := s.Get(ctx, "db-password")
	if err != nil {
		t.Fatalf("Get failed: %v", err)
	}
	if got.Key != "db-password" {
		t.Errorf("expected Key 'db-password', got %q", got.Key)
	}
	if got.Value != "secret123" {
		t.Errorf("expected Value 'secret123', got %q", got.Value)
	}
}

func TestInMemorySecretStore_GetNotFound(t *testing.T) {
	s := NewInMemorySecretStore()
	ctx := context.Background()
	_ = s.Init(ctx, Metadata{})

	_, err := s.Get(ctx, "missing")
	if err == nil {
		t.Fatal("expected error for missing key")
	}
}

func TestInMemorySecretStore_BulkGet(t *testing.T) {
	s := NewInMemorySecretStore()
	ctx := context.Background()
	_ = s.Init(ctx, Metadata{})

	s.Put("k1", "v1")
	s.Put("k2", "v2")

	results, err := s.BulkGet(ctx, []string{"k1", "k2"})
	if err != nil {
		t.Fatalf("BulkGet failed: %v", err)
	}
	if len(results) != 2 {
		t.Fatalf("expected 2 results, got %d", len(results))
	}
	if results[0].Value != "v1" || results[1].Value != "v2" {
		t.Errorf("unexpected values: %v, %v", results[0].Value, results[1].Value)
	}
}

func TestInMemorySecretStore_BulkGetMissing(t *testing.T) {
	s := NewInMemorySecretStore()
	ctx := context.Background()
	_ = s.Init(ctx, Metadata{})

	s.Put("k1", "v1")
	_, err := s.BulkGet(ctx, []string{"k1", "missing"})
	if err == nil {
		t.Fatal("expected error for missing key in BulkGet")
	}
}

func TestInMemorySecretStore_Close(t *testing.T) {
	s := NewInMemorySecretStore()
	ctx := context.Background()
	_ = s.Init(ctx, Metadata{})

	if err := s.Close(ctx); err != nil {
		t.Fatalf("Close failed: %v", err)
	}
	if s.Status(ctx) != StatusClosed {
		t.Errorf("expected StatusClosed, got %s", s.Status(ctx))
	}
}

File: regions/system/library/typescript/building-blocks/src/inmemory_secretstore.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { InMemorySecretStore } from './inmemory_secretstore.js';
import { ComponentError } from './errors.js';

describe('InMemorySecretStore', () => {
  let store: InMemorySecretStore;

  beforeEach(() => {
    store = new InMemorySecretStore();
  });

  it('初期状態は uninitialized', async () => {
    expect(await store.status()).toBe('uninitialized');
  });

  it('init 後は ready になる', async () => {
    await store.init();
    expect(await store.status()).toBe('ready');
  });

  it('close 後は closed になりシークレットがクリアされる', async () => {
    await store.init();
    store.put('k', 'v');
    await store.close();
    expect(await store.status()).toBe('closed');
    await expect(store.getSecret('k')).rejects.toBeInstanceOf(ComponentError);
  });

  it('デフォルト name は inmemory-secretstore', () => {
    expect(store.name).toBe('inmemory-secretstore');
    expect(store.componentType).toBe('secretstore');
  });

  it('コンストラクタで name を指定できる', () => {
    const named = new InMemorySecretStore('custom-secrets');
    expect(named.name).toBe('custom-secrets');
  });

  it('metadata は backend=memory を返す', () => {
    expect(store.metadata()).toEqual({ backend: 'memory' });
  });

  it('put したシークレットを getSecret で取得できる', async () => {
    await store.init();
    store.put('db-password', 'secret123');

    const secret = await store.getSecret('db-password');

    expect(secret.key).toBe('db-password');
    expect(secret.value).toBe('secret123');
  });

  it('存在しないキーの getSecret は ComponentError をスローする', async () => {
    await store.init();
    await expect(store.getSecret('missing')).rejects.toBeInstanceOf(ComponentError);
  });

  it('bulkGet で複数シークレットを取得できる', async () => {
    await store.init();
    store.put('k1', 'v1');
    store.put('k2', 'v2');

    const result = await store.bulkGet(['k1', 'k2']);

    expect(result['k1'].value).toBe('v1');
    expect(result['k2'].value).toBe('v2');
  });

  it('bulkGet でいずれかのキーが存在しない場合は ComponentError をスローする', async () => {
    await store.init();
    store.put('k1', 'v1');

    await expect(store.bulkGet(['k1', 'missing'])).rejects.toBeInstanceOf(ComponentError);
  });

  it('put で既存の値を上書きできる', async () => {
    await store.init();
    store.put('key', 'old');
    store.put('key', 'new');

    const secret = await store.getSecret('key');
    expect(secret.value).toBe('new');
  });
});
